File: 0091-decode-ways/0091-decode-ways.cpp
class Solution {
public:
    int Choices(string s, int index, vector<int> &dp){
        if(index == s.size()){
            return 1;
        }
        if(dp[index] != -1){
            return dp[index];
        }
        int res = 0;
        if(s[index] != '0'){
            res += Choices(s,index+1,dp);
        }
        if(index + 1 <s.size() and (s[index] == '1' or (s[index] == '2' and s[index+1] <= '6'))){
            res += Choices(s, index+2,dp);

        }
        return dp[index] = res;
    }
    int numDecodings(string s) {
        int n = s.length();
        vector<int> dp(n+1,-1);
        return Choices(s,0,dp);
    }
};

File: 0091-decode-ways/0091-decode-ways.js
/**
 * @param {string} s
 * @return {number}
 */

function decoding(s,i){
    if(s[i] === '0'){
        return 0;
    }
    if(i >= s.length - 1){
        return 1;
    }
    let res = decoding(s,i+1);
    if((s[i] === '1' && i<s.length -1) || (s[i] === '2' && i<s.length-1 && s[i+1] <'7'))
        res += decoding(s,i+2);
    
    return res;
}
var numDecodings = function(s) {
    return decoding(s,0);
};